File: 0205-isomorphic-strings/0205-isomorphic-strings.kt
class Solution {
    fun isIsomorphic(s: String, t: String): Boolean {
        if (s.length <= 1) return true
        val map = mutableMapOf<Char, Char>()
        for (i in s.indices) {
            val a = s[i]
            val b = t[i]
            if (map.containsKey(a)) {
                if (map.get(a) == b) continue
                else return false
            } else {
                if (!map.containsValue(b)) map[a] = b
                else return false
            }
        }
        return true
    }
}

File: 0205-isomorphic-strings/0205-isomorphic-strings.ex
defmodule Solution do
  @spec is_isomorphic(s :: String.t, t :: String.t) :: boolean
  def is_isomorphic(s, t) do
    ss = to_charlist(s)
    tt = to_charlist(t)
    
    char_map =
        ss
        |> Enum.zip(tt)
        |> Enum.reduce(%{}, fn {sc, tc}, acc -> Map.put_new(acc, sc, tc) end)
        |> Map.new(fn {key, val} -> {val, key} end)
    
    tt
    |> Enum.map(&Map.get(char_map, &1))
    |> then(&(&1 == ss))
  end
  
end